sheets/api.py

```diff
@@ -738,3 +859,190 @@ def process_assignment_spreadsheets(self):
+            ).select_related("product_coupon__coupon")
+            for assignment in product_coupon_assignments:
+                new_status, status_date = assignment_status_map.get_updated_status_and_date(
+                    bulk_assignment_id,
+                    assignment.product_coupon.coupon.coupon_code,
+                    assignment.email,
+                )
+                if (
+                    new_status
+                    and status_date
+                    and new_status != assignment.message_status
+                ):
+                    assignment.message_status = new_status
+                    assignment.message_status_date = status_date
+                    assignment.updated_on = now_in_utc()
+                    updated_assignments[bulk_assignment_id].append(assignment)
+            ProductCouponAssignment.objects.bulk_update(
+                updated_assignments[bulk_assignment_id],
+                fields=["message_status", "message_status_date", "updated_on"],
+            )
+
+            # Set the BulkCouponAssignment to complete if every coupon has been assigned and
+            # all of the coupon messages have been delivered.
+            unsent_assignments_exist = any(
+                assignment.message_status in {None, MAILGUN_FAILED}
+                for assignment in product_coupon_assignments
+            )
+            if (
+                not unsent_assignments_exist
+                and not assignment_status_map.has_unassigned_codes(bulk_assignment_id)
+            ):
+                BulkCouponAssignment.objects.filter(id=bulk_assignment_id).update(
+                    message_delivery_complete=True, updated_on=now_in_utc()
+                )
+
+            # Update delivery dates in Sheet
+            if assignment_status_map.has_new_statuses(bulk_assignment_id):
+                self.update_sheet_with_new_statuses(
+                    assignment_status_map.get_sheet_id(bulk_assignment_id),
+                    assignment_status_map.get_status_date_rows(bulk_assignment_id),
+                )
+
+        return updated_assignments
+
+    def update_incomplete_assignment_message_statuses(self):
+        """
+        Fetches all BulkCouponAssignments that have one or more undelivered coupon assignments, and
+        attempts to update the message status for each in the database and spreadsheet.
+
+        Returns:
+            dict: Bulk assignment ids mapped to a list of all product coupon assignments that were updated that
+                bulk assignment
+        """
+        bulk_assignments = (
+            BulkCouponAssignment.objects.exclude(assignment_sheet_id=None)
+            .exclude(assignments_started_date=None)
+            .filter(message_delivery_complete=False)
+            .order_by("assignments_started_date")
+            .prefetch_related("assignments")
+        )
+        earliest_date = bulk_assignments[0].assignments_started_date
```

Comment:
any chance there might be no bulk assignments returned here?  `bulk_assignments[0]` would raise an IndexError in that case

Reply:
Yup, I'll handle that case